src/rules/extensions.js

```diff
@@ -137,6 +191,56 @@ module.exports = {
       return false;
     }
 
+    function getEsmExtensionReport(node) {
+
+      const esmExtensions = ['.js', '.ts', '.mjs', '.cjs'];
+      esmExtensions.push(...esmExtensions.map((ext) => `/index${ext}`));
```

Comment:
```suggestion
      const esmExtensions = [].concat(
        '.js',
        '.mjs',
        '.cjs',
        esmExtensions.map((ext) => `/index${ext}`)
      );
```
the `.ts` extension should definitely not be here since it's not ESM.

also, the `.js` extension isn't always ESM, only when `type: module` is present, so this might need to be more complex.

Reply:
Your code isn't valid, but I can remove the 'ts' extension

w.r.t typescript extensions, here's the gotcha: in Typescript packages, you have to `import foo from './foo.js'` instead of `import foo from './foo.ts'`. These extensions are the extensions to check to validate that the path is real, but it should still be replacing a valid `.ts` with `.js`. Currently, the rule will actually set the import to '.ts' instead of '.js'. Which needs fixed...

For this option to be robust in the current landscape, we should really accept a mapping of expected fileExtensions, and the extension to use in the source. something like:

```json
{
"import/extensions": ["error", "always", { 
    "esm": {
        "extensionOverride": { 
            "ts": "js", 
            "tsx": "jsx"
        }, 
        "extensions": ["js", "ts", "cjs", "mjs"]
    }}]
}
```
where extensions without overrides are replaced with themselves. What do you think about this.

Reply:
> also, the .js extension isn't always ESM, only when type: module is present, so this might need to be more complex.

I think for this rule, we can assume when someone enables it that they are using ESM for their non `cjs` files. Would a comment in the doc cover this usecase? 

maybe changing the name to 'fullyResolved' would help communicate more accurately what it's fixing

Reply:
relevant issue: https://github.com/microsoft/TypeScript/issues/49083#issuecomment-1125662016

allowing overrides would let the consumers leave 'ts' replacements as 'ts' (deno migrations) or swapping ts for js (esm migration in typescript project)

Reply:
what is invalid about that code? it works perfectly fine.

I understand about the TS-specific quirks but that should come from the TS resolver and not be hardcoded into a rule that might not be running on TS at all.

I do not think we should *ever* assume that someone is using type:module, which is harmful and should be avoided anyways (altho i believe TS's implementation requires it, not everyone uses TS).

Reply:
just running quickly in this browser terminal:

![image](https://user-images.githubusercontent.com/1173416/219488484-4869c7d1-0857-45ed-8a26-9854c7ffa288.png)

the variable you're assigning to is dependent upon itself, which it can't use because it hasn't defined itself yet.

Reply:
rofl ok fair enough, good catch. in that case, let's keep it as is, but still without TS.
```suggestion

      const esmExtensions = ['.js', '.mjs', '.cjs'];
      esmExtensions.push(...esmExtensions.map((ext) => `/index${ext}`));
```
however we'll still need to handle that `.cjs` isn't ESM, and `.js` is only ESM in type:module.

Unless this is "extensions that might exist in ESM files", and then i'm not sure why it wouldn't include wasm and json?

Reply:
Right, thats why im thinking we should change the rule to "fullyResolved" and then allow users to indicate the extensions they care about, as well as whatever mapping they need.

For my usecase (typescript) i have to check for the existence of both .js and .ts, and replace both with .js

Reply:
tbh this really seems like a uniquely typescript use case, and it probably should live in the TS eslint plugin (cc @bradzacher).

Reply:
The philosophy of TS is and always has been that you write the import paths that you want at runtime. TS purposely does not transform import paths at all - it will always emit the exact same string after transpilation.

In fact currently it is an error to attempt to import a file using `.ts` because as far as TS was concerned - there's no "TS runtime", so importing a `.ts` file makes no sense! EG nodejs can't parse or understand TS, so importing a `.ts` file will crash things.

Because of this, when they added ESM support they added module resolution logic in the typechecker so that when TS sees `./foo.[cm]?jsx?`, it will look for the file `./foo.[cm]?tsx?` and will use that.

[TS 5.0 includes a flag](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#allowimportingtsextensions) which makes it so that `import './foo.ts'` will no longer error for environments where they do understand TS syntax (eg deno or some bundlers). In this instance it will mean that TS will look for that exact file+extension.

To summarise:
- `./foo` will cause TS to do the standard node resolution logic, preferring `ts` extensions
  - eg it'll look for `./foo.ts`, then `./foo.tsx`, then `./foo.js`, then `./foo/index.ts`, etc
- `./foo.[cm]?jsx?` will cause TS to first look for the corresponding `./foo.[cm]?tsx?` file first, then look for the originally specified file.
- `./foo.[cm]?tsx?` will cause TS to look for that exact file (with the correct config)

Reply:
@bradzacher ok, so i'm not really sure what that means here.

It really seems like it should be an option to the TS resolver, not to any particular rule, but it's pretty complex :-)